### ISSUE TITLE
Loss of bit reproducibility across runs

### ISSUE THREAD

**Issue by SalahKouhen** [pyqg/pyqg#250]
I see different results after 1 timestep when I run the same model. I am not sure if this is expected floating point shenanigans I have to live with, or a genuine bug. 

Please see this notebook for the full story:

[Notebook](https://github.com/SalahKouhen/pyqgBitRepro/blob/main/Bit%20reproducibility%20across%20runs.ipynb)

**Comment by rabernat** [comment]
Thanks for reporting this result. I agree it should give the same result. We will try to investigate, but, be forewarned, there are not many people working on pyqg right now. We would really appreciate your help if you have time to investigate further!

**Comment by SalahKouhen** [comment]
Thanks for the quick reply! I am willing to help and will investigate if a new idea comes to me. At the moment there are two ideas that I can't rule out but maybe you or someone on the team can. 

The first is it has something to do with cython. Some variable ends up pointing at the wrong location in the memory and that's where the random values come in. Maybe when you restart the kernel you clear the data from memory but some variables stay pointing to that location? If you have a version of the code with a python kernel it would be interesting to see if the same behaviour happened. 

The second is it has to do with floating point arithmetic being done in a slightly different order. The matrix multiplication is being done with parallelisation allowed, the GIL is turned off, so it is impossible to predict which multiplications will be completed first and so what order they will be added. Since order of addition matters in floating point this will give a different result. The problem with that explanation is surely it would mean the result would be different every time the model runs, not only if you reset the kernel in-between runs.  

I want to use pyqg to look at error growth so resolving (or at least understanding) this issue is key for me.

**Comment by francispoulin** [comment]
@SalahKouhen , thanks for sharing your results.  Any chance you still have the information from so many years ago as to what flags helped?  

It would be nice/esssential to make the results reproducible for sure.

**Comment by SalahKouhen** [comment]
Absolutely. No promises this still works but I added this to the kernel:

IF PYQG_USE_PYFFTW:
    import pyfftw
    pyfftw.interfaces.cache.enable()
ELSE:
    import numpy.fft as npfft
    warnings.warn('No pyfftw detected. Using numpy.fft')
    
#### SALAH ####
# The wisdom file ensures fft done the same way each time, this ensures bit reproducibility
# At around line 230 another bit of code to load up the wisdom. 

#cdef int UseWisdom

#IF UseWisdom==1: 
pyfftw.forget_wisdom()
try:
    wisdom = np.load("wisdom.npy")
except FileNotFoundError:
    print("Wisdom does not exist")
else:
    pyfftw.import_wisdom(wisdom)
    print("Wisdom loaded")
        
#### SALAH END ####

I added this to the end of init:

            #### SALAH ####
            #IF UseWisdom==1: 
            try:
                wisdom = np.load("wisdom.npy")
            except FileNotFoundError:
                np.save('wisdom', pyfftw.export_wisdom())
                print("Wisdom saved")      
            else:
                print("Wisdom not saved as old wisdom found")
            #### SALAH END ####

Hope that helps.
